### PR TITLE
Hoc loading error states

### LINKED ISSUE #1248: Create loading page
**Is your feature request related to a problem? Please describe.**
Our current loading state for second page load is a blank page with the text `Loading...`.  We should create a loading screen that maintains much of the page furniture (&, depending on UX requirements, may render a "skeleton screen" or similar preview.

Adding a smoother transition to this screen is covered in https://github.com/bbc/simorgh/issues/1302 which doesnt block this.

Blocked on UX's specifications for load page appearance.

**Describe the solution you'd like**
This should follow https://bbc.github.io/gel/foundations/routing/#awaiting-content

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

**Additional context**

- [x] Initially labelled with ["Refinement needed"](https://github.com/bbc/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by chris-hinds**
Resolves #1248 

**Overall change:**

This PR branches from @sareh original MVP loading state, with the intention of making the Loading and error states more re-useable when adding different page types.

This is done by using HOC (Higher Order Components) to provide a loading and an error state. This enables to reduce of a lot of complexity in the article container and in the future homepage container, removing the need for these to know about error and loading states.

It also introduces a `PageWrapper` that is used to load the necessary page furniture which should be fairly similar if not the same across articles and homepages.

I would want to refactor this some more at a later date by producing a Layout that then wraps each page type container, again reducing the required logic in both the article and the homepage containers.

**Code changes:**

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [x] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/App/App.jsx
Comment (chris-hinds) [inline src/app/containers/App/App.jsx]: Oops ignore this for now, this will be removed, but was refactoring the main App component to use Hooks.

File: src/app/containers/Article/index.jsx
Comment (dr3) [inline src/app/containers/Article/index.jsx]: If theres an error in say, `PageWrapper`, the try catch would catch it, but then the `catch` would just cause the error again

Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: Good points, so Haven't looked at this part of the code too much yet, was just focussing on the original loading and error state. But we could have a WithData HOC that handles this same try catch logic here?

Comment (dr3) [reply src/app/containers/Article/index.jsx]: We can remove the try catch here, and use Error Boundaries in App to toggle the `error` state, therefor showing the error page

File: src/app/containers/Article/index.jsx
Comment (sareh) [inline src/app/containers/Article/index.jsx]: Reading up on the recompose library it seems that the author recommends not using it for new projects. 
https://github.com/acdlite/recompose#a-note-from-the-author-acdlite-oct-25-2018 
https://github.com/acdlite/recompose/issues/756#issuecomment-438674573

Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: Hmm so I read that more as using recompose is still fine, it will not receive new features etc but that there is now an alternative.

I mentioned in the top comment about moving to hooks in the near future. I just wonder if it would be worth finishing up this sort of approach, given we need this functionality relatively soon.

And then refactoring again as this is an MVP.

Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: but I am going to try an alternative today, and try and estimate the level of work involved.

File: src/app/containers/PageHandlers/withLoading.jsx
Comment (chris-hinds) [inline src/app/containers/PageHandlers/withLoading.jsx:9]: This ideally would be a separate component that defines the loading page layout. This would be re-factored when we get some official UX input

File: src/app/containers/PageHandlers/withPageWrapper.jsx
Comment (chris-hinds) [inline src/app/containers/PageHandlers/withPageWrapper.jsx]: With pageWrapper you would be able to pass the in the Page Layout that you would like. So we could have a different layout for topic pages for instance. And these layouts would be defined in the new Layouts directory making it very easy to compose a page type with a different layout.

File: src/app/containers/Article/index.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: Actually found this to be a really good read when looking into HOC vs Hooks https://medium.com/@albertchu539/higher-order-components-in-a-react-hooks-world-69fe1f0b0791

I think HOC's and something like `recompose` used here still has its place.

The hooks alternative mentioned by the recompose author is best used when your HOC's were just being used to enhance your base component with props or data derived from state and not actually when you want to render an entirely different component as we do here.

Going without something like recompose would reduce the modularity of these error handling and loading HOC's as we would still have to implement the same if/else logic in each container that wishes to use default error and loading states.

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx:36]: These Error mains can likely be consolidated to a single call, so if data is bad, or status != 200 or the try catch is thrown, it falls to the error. 

Also I think status is undefined here, but I haven't run it

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx]: Gah ES5, RUN! 🙈 :P

File: src/app/containers/Article/index.jsx
Comment (dr3) [reply src/app/containers/Article/index.jsx]: As its advised not to use recompose, and I doubt maintainance will be high on anyones list, we should pull out this teeny function into simorgh helpers, with credit ofcourse :P https://github.com/acdlite/recompose/blob/master/src/packages/recompose/compose.js

File: src/app/containers/Article/index.jsx
Comment (dr3) [inline src/app/containers/Article/index.jsx:24]: 😍

File: src/app/containers/PageHandlers/withError.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withError.jsx]: I was gunna say i have a personal preference that we should flip this, but Im disagreeing with myself now :P

**Review by dr3 — CHANGES_REQUESTED** [review]
Loveee this, so clean and simple and modular 😍 

Also this PR is draft, so soz for the review but its so close to being great 🙈

File: src/app/containers/Article/index.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: Unfortunately from the looks of things react error boundaries do not yet work for SSR applications.

The lifecycle methods used for reacts error boundaries e.g. `componentDidCatch` are not yet supported when using `renderToString` or `renderToStaticMarkup`

Comment (dr3) [reply src/app/containers/Article/index.jsx]: 😢

File: src/app/containers/PageHandlers/withData.jsx
Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx:36]: ahh good spot on status being undefined, although I think the function requires an return statement at the end, so we either return `null` or simply return the error component with a status code of 500. 

Thoughts?

Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx:36]: Confirmed, assume data is `null` the final return statement is needed otherwise nothing renders.

Comment (dr3) [reply src/app/containers/PageHandlers/withData.jsx:36]: I think you could do this, cant see any obvious issues, but ive not run it 🙈 😂 
```
const WithData = Component => ({ data }) => {
  let status;

  if (data) {
    const { data: articleData } = data;
    status = data.status; // eslint-disable-line prefer-destructuring

    if (status && articleData && status === 200) {
      try {
        return <Component articleData={articleData} />
      } catch (err) {
        logger.error('Oh noos' + err);
      }
    }
  };

  return <ErrorMain status={status || 500} />;
};
```

Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx:36]: If the data prop was null or anything other than an object this would fail with `Cannot read property 'data' of null`.

Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx:36]: Although we change the default prop for data to be an empty object?

Comment (dr3) [reply src/app/containers/PageHandlers/withData.jsx:36]: Ermm whoops, yep I was wrong, data should be required and have no default, updated code

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx]: [`The let statement declares a block scope local variable`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) so its only setting the value inside `if (data) {` and not the one on line 10 :(

Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx]: 🙈

File: src/app/containers/Article/index.jsx
Comment (rossgaskell) [inline src/app/containers/Article/index.jsx]: Really pedantic, but don't have much else to add 😂 

Would it be better for the HOCs to be named with Lower Camel Case or is there a reason we shouldn't be doing it?

File: src/app/containers/PageHandlers/withData.jsx
Comment (rossgaskell) [inline src/app/containers/PageHandlers/withData.jsx]: Would we want to be defining the default status code in ErrorMain itself? There is already a case of this in ErrorMain itself: https://github.com/bbc/simorgh/blob/37ca772ed0f6a8630693578fcd3fc4814e90fe79/src/app/containers/ErrorMain/index.jsx#L35

**Review by rossgaskell — DISMISSED** [review]
Looking fabulous!

**Review by rossgaskell — COMMENTED** [review]
Looking fabulous!

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [reply src/app/containers/PageHandlers/withData.jsx]: Im personally a fan of being explicit of the type of error we should show, rather than relying on the defaults, however I see your point Ross :)

File: package.json
Comment (dr3) [inline package.json]: As its advised not to use recompose, and I doubt maintainance will be high on anyones list, we should pull out this teeny function into simorgh helpers, with credit ofcourse :P https://github.com/acdlite/recompose/blob/master/src/packages/recompose/compose.js

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/containers/Article/index.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: Nope, I think you are right, I have really followed the naming convention here. Will update.

**Comment by jamesbhobbs** [comment]
PR closed and reopened to trigger Travis PR build

File: package.json
Comment (chris-hinds) [reply package.json]: done.

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx]: I think these prop types should be inline with the component, not inside it

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx:51]: Should this be `node` instead?

File: src/app/containers/PageHandlers/withData.jsx
Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx]:
```const WithData = Component => {
  const DataContainer = ({ data, ...props }) => {
    const { hasDataAnd200Status, status, assetData } = shouldRender(data);
    if (hasDataAnd200Status) {
      try {
        return <Component assetData={assetData} {...props} />;
      } catch (err) {
        logger.error(err);
      }
    }

    return <ErrorMain status={status} />;
  };

  DataContainer.propTypes = {
    data: shape(articlePropTypes),
  };

  DataContainer.defaultProps = {
    data: shape,
  };

  return DataContainer;
};
```

Is this what you are suggesting @dr3

Comment (dr3) [reply src/app/containers/PageHandlers/withData.jsx]: Yeah, you shouldnt reference DataContainer from inside itself :)

File: src/app/containers/PageHandlers/__snapshots__/withData.test.jsx.snap
Comment (rossgaskell) [inline src/app/containers/PageHandlers/__snapshots__/withData.test.jsx.snap:21]: Is this the expected output for this test?

File: src/app/containers/PageHandlers/__snapshots__/withError.test.jsx.snap
Comment (rossgaskell) [inline src/app/containers/PageHandlers/__snapshots__/withError.test.jsx.snap]: Is this the expected output for this test?

File: src/app/containers/PageHandlers/withData.jsx
Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx]: yeah, good spot @dr3

File: src/app/containers/PageHandlers/__snapshots__/withError.test.jsx.snap
Comment (chris-hinds) [reply src/app/containers/PageHandlers/__snapshots__/withError.test.jsx.snap]: Good spot, I guess technically it is as I am not passing the correct props in the test :D Updating now.

File: src/app/containers/PageHandlers/__snapshots__/withData.test.jsx.snap
Comment (chris-hinds) [reply src/app/containers/PageHandlers/__snapshots__/withData.test.jsx.snap:21]: ahhh, against incorrect props in the test data. 🙈

File: src/app/containers/PageHandlers/withData.jsx
Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx:51]: Node is a bit too wide open for my liking, it allows you to pass through any renderable item, arrays, strings, numbers , elements etc. Where as element species a single React element.

https://reactjs.org/docs/typechecking-with-proptypes.html#proptypes

**Comment by chris-hinds** [comment]
@dr3 & @rossgaskell have made some updates based on your comments.

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (dr3) [inline src/app/Layouts/defaultPageWrapper.jsx]: I dont think this is right, its left as null https://github.com/bbc/simorgh/blob/latest/src/app/containers/Article/index.jsx#L72 as the logic of the defaulting is handled in https://github.com/bbc/simorgh/blob/latest/src/app/contexts/RequestContext/getOriginContext.js

Adding the default here, the client will always think its on `.co.uk`

File: src/app/containers/Article/index.jsx
Comment (dr3) [inline src/app/containers/Article/index.jsx]: We should get rid of ArticleMain and just use this container

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx]: 😱 it defaults to a prop-type

File: src/app/containers/PageHandlers/withPageWrapper.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withPageWrapper.jsx:21]:
```suggestion
  };

```

File: src/app/containers/PageHandlers/withPageWrapper.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withPageWrapper.jsx]:
```suggestion
```

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (dr3) [inline src/app/Layouts/defaultPageWrapper.jsx]: ~This should probably go in the article container~

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: Oops, yup. fixing.

File: src/app/containers/PageHandlers/withData.jsx
Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx]: ahh nope, this shouldn't have been left in

File: src/app/containers/PageHandlers/__snapshots__/withPageWrapper.test.jsx.snap
Comment (dr3) [inline src/app/containers/PageHandlers/__snapshots__/withPageWrapper.test.jsx.snap:3]: Hmm, this snapshot doesn't show the `<h1>Hola</h1>` anywhere 🤔

File: src/app/containers/Article/index.test.jsx
Comment (dr3) [inline src/app/containers/Article/index.test.jsx:1]: Im dubious of this testing file, it feels like it should stub all the `PageHandlers`, as its really testing them

File: src/app/containers/PageHandlers/__snapshots__/withPageWrapper.test.jsx.snap
Comment (chris-hinds) [reply src/app/containers/PageHandlers/__snapshots__/withPageWrapper.test.jsx.snap:3]: This is an issue when using shallowRender. Will update this to use the full render

File: src/app/containers/Article/index.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.jsx]: I think we should do this in another refactor, this is is already quite long as is. Will create an issue to do this.

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (dr3) [reply src/app/Layouts/defaultPageWrapper.jsx]: Nope, ignore me, its fine here for now, eventually it will have a different, general, path

Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: Service context provider will be needed across pageTypes and help needs the values from it.

Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: haha ignored :p

Comment (dr3) [reply src/app/Layouts/defaultPageWrapper.jsx]: I was pointing at the helmet, for the manifest json, but nvm :P

Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: Yup, but helmet requires service, which it gets from the service context.

File: src/app/containers/Article/index.test.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.test.jsx:1]: Will look at this again, but I think this should change under another issue. For now the fully rendered snapshot is effect testing the composing of the multiple HOC's.

File: src/app/containers/Article/index.test.jsx
Comment (dr3) [inline src/app/containers/Article/index.test.jsx]: Still dubious of these tests, as really they're testing the PageHanders again

Would be interested in others opinions, @pjlee11 @sareh

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx]:
```suggestion
  status: deepGet(['status'], data) || 500,
```

File: src/app/containers/PageHandlers/withData.jsx
Comment (dr3) [inline src/app/containers/PageHandlers/withData.jsx]:
```suggestion
  if (assetData && status === 200) {
    hasDataAnd200Status = true;
  }
```

**Review by dr3 — CHANGES_REQUESTED** [review]
So close, just a couple things before 👍

File: src/app/containers/PageHandlers/withData.jsx
Comment (chris-hinds) [reply src/app/containers/PageHandlers/withData.jsx]: Any particular reason for this or is it just preference? This just seems like a cleaner way to simply set a variable.

Comment (dr3) [reply src/app/containers/PageHandlers/withData.jsx]: Cleaner would be 
```
hasDataAnd200Status = assetData && status === 200
```
Im not a fan of if statements without a block, theyre so rare they just tend to be confusing imo

File: src/app/containers/Article/index.test.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.test.jsx]: Going to look at removing the loading state test from here and mocking the HOC's

File: src/app/containers/Article/index.test.jsx
Comment (chris-hinds) [reply src/app/containers/Article/index.test.jsx:1]: This has now been updated to mock each of the HOC (the page handlers) the idea behind this test is to ensure that the each HOC is called in the correct order ultimately returning the original components passed in.

The article itself is tested within ArticleMain and each HOC is tested individually as well. This is just testing the composing of the HOC.

Any further testing/refactoring will be moved to a new issue.

**Review by rossgaskell — DISMISSED** [review]
image

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (dr3) [inline src/app/Layouts/defaultPageWrapper.jsx]: This needs to be optional everywhere, defaulting to `null`
```
Warning: Failed prop type: The prop `bbcOrigin` is marked as required in `PageWrapper`, but its value is `null`.
```

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: Ohh did not realise that, changing now

**Review by rossgaskell — DISMISSED** [review]
image

File: src/app/Layouts/defaultPageWrapper.jsx
Comment (chris-hinds) [reply src/app/Layouts/defaultPageWrapper.jsx]: updated.

**Review by dr3 — COMMENTED** [review]
LGTM

Im not 100% on [these tests](https://github.com/bbc/simorgh/pull/1589/files?file-filters%5B%5D=.js&file-filters%5B%5D=.jsx#diff-9e1ed69f2ef64e4fa708890f683b3e97) though, so I would like a review from another code owner. Holding my review until then

**Review by jamesdonoh — DISMISSED** [review]
The tests as they stand provide enough coverage against breakage, let's raise a separate issue to improve them further if needed - this work is a blocker for others and we need to finish it up. Thanks @hindsc52 for taking this on and everyone for their reviews.

**Review by dr3 — DISMISSED** [review]
lgtm

**Comment by jamesdonoh** [comment]
Blocked by https://github.com/bbc/simorgh/issues/1661

**Comment by dr3** [comment]
👋 When this PR is ready, it will need a bump and changelog update :) Details can be found [here](https://github.com/bbc/simorgh/blob/a75ad1fbf4a720bf6147d7f57ee3d8e3a7337632/CONTRIBUTING.md#versioning-and-changelog)

**Comment by chris-hinds** [comment]
Please note the current CC issues are for tests so can be ignored.

**Review by rossgaskell — APPROVED** [review]
image
^^ It definitely is this time 😂

**Comment by jamesbrumpton** [comment]
LGTM. Happy for this to be merged.

**Pull request merged**
